beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java

```diff
@@ -0,0 +1,171 @@
+    @Getter private final boolean startInclusive;
+    @Getter @Nullable private Offset endOffsetInclusive;
+    // weather to disallow adding claim past endOffset
+    @Getter private boolean extensible;
+    private transient boolean finished = false;
+
+    private OffsetRange(@Nullable Offset start, boolean inclusive) {
+      this.startOffset = start;
+      this.startInclusive = inclusive;
+      extensible = true;
+    }
+
+    private OffsetRange() {
+      this(null, false);
+    }
+
+    private OffsetRange(Offset startExclusive) {
+      this(Objects.requireNonNull(startExclusive), false);
+    }
+
+    public OffsetRange(Offset startOffsetExclusive, Offset endOffsetInclusive) {
+      this.startOffset = Objects.requireNonNull(startOffsetExclusive);
+      this.startInclusive = false;
+      this.endOffsetInclusive = Objects.requireNonNull(endOffsetInclusive);
+      this.extensible = false;
+    }
+
+    boolean getAndCommitClaim(Offset offset) {
+      if (!extensible) {
+        if (finished) {
+          return false;
+        }
+        finished = offset.equals(endOffsetInclusive);
+        return true;
+      }
+      endOffsetInclusive = offset;
+      return true;
+    }
+
+    @Nullable
+    public OffsetRange asPrimary() {
+      if (endOffsetInclusive == null) {
+        return null;
+      }
+      return new OffsetRange(startOffset, endOffsetInclusive);
+    }
+
+    @Nullable
+    public OffsetRange asResidual() {
+      if (endOffsetInclusive == null) {
+        return null;
+      }
+      return new OffsetRange(endOffsetInclusive);
+    }
+
+    public boolean isInitial() {
+      return startOffset == endOffsetInclusive && endOffsetInclusive == null;
+    }
+
+    void terminate() {
+      this.extensible = false;
+      this.finished = true;
+    }
+
+    @Override
+    public OffsetRestrictionTracker newTracker() {
+      Preconditions.checkState(
+          !finished, "Tracker can be created only for not-yet split restrictions");
+      Preconditions.checkState(
+          extensible, "Tracker can be created only for not-yet split restrictions");
+      return new OffsetRestrictionTracker(this);
+    }
+  }
+
+  private final OffsetRange currentRestriction;
+
+  private OffsetRestrictionTracker(OffsetRange initial) {
+    this.currentRestriction = initial;
+  }
+
+  @Override
+  public boolean tryClaim(Offset position) {
+    return currentRestriction.getAndCommitClaim(position);
+  }
+
+  @Override
+  public OffsetRange currentRestriction() {
+    return currentRestriction;
+  }
+
+  @Override
+  public @Nullable SplitResult<OffsetRange> trySplit(double fractionOfRemainder) {
+    @Nullable OffsetRange primary = currentRestriction.asPrimary();
+    @Nullable OffsetRange residual = currentRestriction.asResidual();
+    if (residual != null) {
+      currentRestriction.terminate();
+    }
+    return SplitResult.of(primary, MoreObjects.firstNonNull(residual, currentRestriction));
```

Comment:
This is construct is hard to understand. Can we have something more descriptive? eg:
```
    @Nullable OffsetRange primary = currentRestriction.asPrimary(); 
    if (primary == null) {
      // We can not checkpoint current restriction.
      currentRestriction.terminate();
      return SplitResult.of(null, currentRestriction);
    }
    return SplitResult.of(primary, Objects.requireNonNull(currentRestriction.asResidual()));
```

Reply:
changed that to
```java
    if (currentRestriction.isSplittable()) {
      currentRestriction.terminate();
      return SplitResult.of(currentRestriction.asPrimary(), currentRestriction.asResidual())
    }
    return SplitResult.of(null, currentRestriction);
```